src/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs

```diff
@@ -1908,16 +1915,21 @@ public bool IsCompleted
                 }
             }
 
-            internal unsafe DisconnectAsyncResult(HttpListener httpListener, ulong connectionId)
+            internal DisconnectAsyncResult(HttpListener httpListener, ulong connectionId)
             {
                 if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"HttpListener: {httpListener}, ConnectionId: {connectionId}");
                 _ownershipState = 1;
                 _httpListener = httpListener;
                 _connectionId = connectionId;
 
-                // we can call the Unsafe API here, we won't ever call user code
-                _nativeOverlapped = httpListener.RequestQueueBoundHandle.AllocateNativeOverlapped(s_IOCallback, state: this, pinData: null);
-                if (NetEventSource.IsEnabled) NetEventSource.Info($"DisconnectAsyncResult: ThreadPoolBoundHandle.AllocateNativeOverlapped({httpListener._requestQueueBoundHandle}) -> {_nativeOverlapped->GetHashCode()}");
+                // HttpListener.Stop() will reset its RequestQueueBoundHandle. Take a reference to the one we
+                // allocate on to ensure the OVERLAPPED structure is freed on the proper request queue if the
+                // disconnect I/O completion comes in after Stop() is called.
+                _disconnectRequestQueue = httpListener.RequestQueueBoundHandle;
+
+                // We can call the Unsafe API here, we won't ever call user code.
+                _nativeOverlapped = _disconnectRequestQueue.AllocateNativeOverlapped(s_IOCallback, state: this, pinData: null);
+                if (NetEventSource.IsEnabled) NetEventSource.Info($"DisconnectAsyncResult: ThreadPoolBoundHandle.AllocateNativeOverlapped({_disconnectRequestQueue}) -> {_nativeOverlapped->GetHashCode()}");
```

Comment:
```suggestion
                if (NetEventSource.IsEnabled) NetEventSource.Info($"DisconnectAsyncResult: {_disconnectRequestQueue}.AllocateNativeOverlapped -> {_nativeOverlapped->GetHashCode()}");
```

Reply:
My change here is only to reference `_disconnectRequestQueue` versus `httpListener.RequestQueueBoundHandle`. I don't want to change the message; it's not clear to me if that whatever `ThreadPoolBoundHandle.ToString()` does will lead to a sensible message.

Reply:
Ok.